src/lib/operators/insert.cpp

```diff
@@ -107,136 +93,132 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
+      }
 
-    auto remaining_rows = total_rows_to_insert;
-    while (remaining_rows > 0) {
-      auto current_chunk = _target_table->get_chunk(static_cast<ChunkID>(_target_table->chunk_count() - 1));
-      auto rows_to_insert_this_loop = std::min(_target_table->max_chunk_size() - current_chunk->size(), remaining_rows);
+      const auto target_chunk_num_inserted_rows =
+          std::min<size_t>(_target_table->max_chunk_size() - target_chunk->size(), remaining_rows);
 
-      // Resize MVCC vectors.
-      current_chunk->get_scoped_mvcc_data_lock()->grow_by(rows_to_insert_this_loop, MvccData::MAX_COMMIT_ID);
+      _target_chunk_ranges.emplace_back(
+          ChunkRange{target_chunk_id, target_chunk->size(),
+                     static_cast<ChunkOffset>(target_chunk->size() + target_chunk_num_inserted_rows)});
 
-      // Resize current chunk to full size.
-      auto old_size = current_chunk->size();
-      for (ColumnID column_id{0}; column_id < current_chunk->column_count(); ++column_id) {
-        typed_segment_processors[column_id]->resize_vector(current_chunk->get_segment(column_id),
-                                                           old_size + rows_to_insert_this_loop);
-      }
+      // Grow MVCC columns while simultaneously locking them for the current transaction.
+      target_chunk->get_scoped_mvcc_data_lock()->grow_by(target_chunk_num_inserted_rows, context->transaction_id());
 
-      remaining_rows -= rows_to_insert_this_loop;
+      // Grow data Segments.
+      auto old_size = target_chunk->size();
+      for (ColumnID column_id{0}; column_id < target_chunk->column_count(); ++column_id) {
+        resolve_data_type(_target_table->column_data_type(column_id), [&](const auto data_type_t) {
+          using ColumnDataType = typename decltype(data_type_t)::type;
 
-      // Create new chunk if necessary.
-      if (remaining_rows > 0) {
-        _target_table->append_mutable_chunk();
-        end_chunk_id++;
+          resize_value_segment<ColumnDataType>(target_chunk->get_segment(column_id),
+                                               old_size + target_chunk_num_inserted_rows);
+        });
       }
+
+      remaining_rows -= target_chunk_num_inserted_rows;
     }
   }
-  // TODO(all): make compress chunk thread-safe; if it gets called here by another thread, things will likely break.
 
-  // Then, actually insert the data.
-  auto input_offset = 0u;
-  auto source_chunk_id = ChunkID{0};
-  auto source_chunk_start_index = 0u;
+  /**
+   * 2. Insert the Data into the memory allocated in the first step. Write the transaction_context's transaction_id into
+   *    all allocated rows.
+   */
+  auto source_row_id = RowID{ChunkID{0}, ChunkOffset{0}};
 
-  for (auto target_chunk_id = start_chunk_id; target_chunk_id < end_chunk_id; target_chunk_id++) {
-    auto target_chunk = _target_table->get_chunk(target_chunk_id);
+  for (const auto& target_chunk_range : _target_chunk_ranges) {
```

Comment:
This loop feels unintuitive to me. I wouldn't have expected a loop over the chunks that we insert into as the outermost loop. Also, resolving the data type for each value seems to be quite expensive. Isn't it comparable to using AllTypeVariants?

In pseudocode, this seems to be easier:

```
next_target_chunk_range = target_chunk_ranges[0]
next_target_chunk_offset = next_target_chunk_range.begin
foreach column:
  foreach input_chunk:
    iterate_segment<column.type>(input_chunk.get_segment(column), input_it):
      segment = static_cast<ValueSegment<...>>(next_target_chunk_range.chunk.get_segment(column))
      segment.values[next_target_chunk_offset] = *input_it
      next_target_chunk_offset++
      if (next_target_chunk_offset > next_target_chunk_range.end):
        // update next_target_chunk_range
```

Reply:
Gods, I am not resolving the data type for every value - just for every target chunk range, covering hundred thousands of rows!

Reply:
I am certain there are other ways to structure the loops. The way I do it is that I first build the target-chunk ranges (aka "this is where the inserted data goes") and then I fill these ranges in the seconds step. I think it's okay enough.

Reply:
Oups. Might have misread that. That makes the performance aspect void, but I'd still like to discuss the general approach from a readability perspective.

Reply:
Didn't see the second comment. Let's work through the rest and I'll have another look at this during my second pass.

Reply:
I tried quite a while to get my version to work. Turns out your approach is better suited to handle self-inserts.